pacman/model/routing_table_by_partition/multicast_routing_table_by_partition.py

```diff
@@ -13,39 +13,68 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+from pacman.model.graphs.application import ApplicationVertex
+from pacman.exceptions import PacmanInvalidParameterException
+import logging
+
+log = logging.getLogger(__name__)
+
 
 class MulticastRoutingTableByPartition(object):
     """ A set of multicast routing path objects
     """
 
     __slots__ = [
-        # dict mapping (x,y) -> dict mapping (partition) -> routing table entry
+        # dict mapping (x,y) -> dict mapping (source_vertex, partition_id))
+        # -> routing table entry
         "_router_to_entries_map"
     ]
 
     def __init__(self):
         self._router_to_entries_map = dict()
 
-    def add_path_entry(self, entry, router_x, router_y, partition):
+    def add_path_entry(
+            self, entry, router_x, router_y, source_vertex, partition_id):
         """ Adds a multicast routing path entry
 
         :param MulticastRoutingTableByPartitionEntry entry: the entry to add
         :param int router_x: the x coord of the router
         :param int router_y: the y coord of the router
-        :param AbstractSingleSourcePartition partition:\
-            the partition containing the machine edge
+        :param source_vertex: The source that will send via this entry
+        :type source_vertex: ApplicationVertex or MachineVertex
+        :param str partition_id: The id of the partition being sent
         """
 
         # update router_to_entries_map
         key = (router_x, router_y)
-        if key not in self._router_to_entries_map:
-            self._router_to_entries_map[key] = dict()
+        entries = self._router_to_entries_map.get(key)
+        if entries is None:
+            entries = dict()
+            self._router_to_entries_map[key] = entries
+
+        if isinstance(source_vertex, ApplicationVertex):
```

Comment:
It is OK to have an app vertex already there and then add it again, and similarly for machine vertex; this is handled by merge code later on.  It is not OK to have both app vertex and any machine vertices though.  Both of those are checked; if not an app vertex, it is assumed to be a machine vertex (it could check but then gigo).